### PR TITLE
Mysql FluentD plugin for OMS agent

### PR DESCRIPTION
**Pull request by NarineM**
1. Added a fluentd plugin for MySQL server
2. Added unit test for MySQL plugin
3. Added mysql2 ruby API gem  and updated ruby.data to install mysql2 gem under the agent's local ruby path.  Re-arranged ruby.data content for existing gems based on the genRubyInstaller.sh results to help to track only new changes moving forward instead of only adding diff related to MySQL2 gem.  Will send the fix for genRubyInstaller.sh in a separate pull request.
4. Added fluentd conf configuration file that is installed only under sysconf path. 
5 .Added quick documentation how to configure MySQL plugin and pre-reqs. Will need to update existing MySQL omi provider related sections after omi provider is removed. This is just a temp quick setup guide that will need to be removed later on.

@Microsoft/omsagent-devs @robbiezhang

### CODE REVIEW

File: docs/How to setup mysql plugin.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# How to setup Mysql plugin for OMS agent
```
Comment (robbiezhang) [inline docs/How to setup mysql plugin.md]: Don't use space in the file name. Use '-' instead.

File: installer/conf/omsagent.d/mysqlworkload.conf [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# This file is an example of what the config file pushed by the server should look like
```
Comment (robbiezhang) [inline installer/conf/omsagent.d/mysqlworkload.conf]: @agup006, shall we unify the format of the workload conf name?

File: installer/datafiles/base_omsagent.data [hunk 3]
```diff
@@ -32,7 +32,7 @@ MAINTAINER:              'Microsoft Corporation'
 /etc/opt/microsoft/omsagent/sysconf/omsagent.d/collectd.conf;           installer/conf/omsagent.d/collectd.conf;               644; root; root
 /etc/opt/microsoft/omsagent/sysconf/omsagent.d/statsd.conf;             installer/conf/omsagent.d/statsd.conf;                 644; root; root
 /etc/opt/microsoft/omsagent/sysconf/omsagent.d/auditlog.conf;           installer/conf/omsagent.d/auditlog.conf;               644; root; root
-/etc/opt/microsoft/omsagent/sysconf/omsagent.d/mongostat.conf;          installer/conf/omsagent.d/mongostat.conf;              644; root; root
```
Comment (robbiezhang) [inline installer/datafiles/base_omsagent.data]: please follow the GIT best practice, e.g. run git rebase master before merge

File: source/code/plugins/in_mysql_workload.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
+module Fluent
```
Comment (robbiezhang) [inline source/code/plugins/in_mysql_workload.rb]: @agup006, shall we unify the name of plugin for workloads?

File: source/code/plugins/in_mysql_workload.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  class MysqlWorkloadInput < Fluent::Input
+    Plugin.register_input('mysql_workload', self)
+
+    # Define `router` method to support v0.10.57 or earlier
```
Comment (robbiezhang) [inline source/code/plugins/in_mysql_workload.rb]: omsagent references fluentd 0.12.24. You don't need to support v0.10.57 or earlier.

File: source/code/plugins/in_mysql_workload.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    def start
+      @thread = Thread.new(&method(:run))
+    end
+
+    def shutdown
+      Thread.kill(@thread)
+    end
+
+    def run
+      @mysql_lib = MysqlWorkload_Lib.new(@host, @port, @username, @password, @database, @encoding, OMS::Log)
```
Comment (robbiezhang) [inline source/code/plugins/in_mysql_workload.rb]: you need to require_relative 'omslog' and parse in @log here.

File: source/code/plugins/in_mysql_workload.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    config_param :database, :string, :default => nil
+    config_param :encoding, :string, :default => 'utf8'
+    config_param :interval, :time, :default => '1m'
+    config_param :tag, :string
+
+    def configure(conf)
+      super
+    end
+
+    def start
```
Comment (robbiezhang) [inline source/code/plugins/in_mysql_workload.rb]: call super at the beginning

File: source/code/plugins/in_mysql_workload.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+      @thread = Thread.new(&method(:run))
+    end
+
+    def shutdown
```
Comment (robbiezhang) [inline source/code/plugins/in_mysql_workload.rb]: call super at the beginning

File: source/code/plugins/in_mysql_workload.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      Thread.kill(@thread)
```
Comment (robbiezhang) [inline source/code/plugins/in_mysql_workload.rb]: do you need to close the mysql connection and release resources before leave?

File: source/code/plugins/mysql_workload_lib.rb [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+  end
+
+  def get_mysql_hostname
+    query("SHOW VARIABLES LIKE 'hostname'").each do |row|
+      return row.fetch('Value')
+    end
+  end
+
+  def transform_row(row)
+    # creates a new hash table where keys from the input row are replaced with the Variable_name key name, and the value of the key Variable_name
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: please break the comments into multiple line, it's hard to review in Git

File: source/code/plugins/mysql_workload_lib.rb [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+        value = row.fetch(key_to_find)
+        break
+      end
+    end
+    return value
+  end
+
+  def get_server_database_type
+    #retrieve engine info from default_storage_engine
+    engine_query = "show variables like 'default_storage_engine'" #some versions of mysql might not populated this field so use default
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: comments is out of boundary in Git

File: source/code/plugins/mysql_workload_lib.rb [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+    engine_info = query(engine_query)
+    engine_info_value = "InnoDB" # default
+    engine_info.each do |row|
+      engine_info_value = row["Value"]
+    end
+    return engine_info_value
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: always return the last one?

File: source/code/plugins/mysql_workload_lib.rb [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+      end
+    else
+      return @mock_interface.query(input_query)
+    end
+
+  end
+
+  def get_mysql_hostname
+    query("SHOW VARIABLES LIKE 'hostname'").each do |row|
+      return row.fetch('Value')
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: always return the first one?

File: source/code/plugins/mysql_workload_lib.rb [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+
+    # Add NumConnections
+    num_connections = Hash.new
+    num_connections_value = get_value(global_stats, "Threads_connected")
+    num_connections.store("NumConnections", num_connections_value)
+    final_records.push(num_connections)
+
+    # Add MaxConnections
+    max_connections = Hash.new
+    max_connections_value = get_value(env_var_stats, "max_connections")
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: is the key case sensitive?

File: source/code/plugins/mysql_workload_lib.rb [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+    uptime.store("Uptime", uptime_value)
+    final_records.push(uptime)
+
+    # add KeyCacheHitPct
+    key_cache_hit_pct = Hash.new
+    key_reads = get_value(global_stats,'Key_reads')
+    key_read_requests = get_value(global_stats,'Key_read_requests')
+    value = 0 #initialize
+    if key_read_requests.to_i != 0
+      value = (key_reads.to_i / key_read_requests.to_i) * 100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: it should be to_f (to float), otherwise it will always give you a round integer, and you lose the accuracy before multiple by 100.

File: source/code/plugins/mysql_workload_lib.rb [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+    slow_query_pct.store("SlowQueryPct", queries )
+    final_records.push(slow_query_pct)
+
+    #add KeyCacheWritePct
+    key_cache_write_pct = Hash.new()
+    key_writes = get_value(global_stats,"Key_writes")
+    key_writes_requests = get_value(global_stats,"Key_writes_requests")
+    value = 0
+    if(key_writes_requests.to_i != 0)
+      value = (key_writes.to_i)/(key_writes_requests.to_i) * 100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: same here to_f

File: source/code/plugins/mysql_workload_lib.rb [hunk 17]
```diff
@@ -0,0 +1,442 @@
...
+    key_cache_write_pct.store("KeyCacheWritePct", value)
+    final_records.push(key_cache_write_pct)
+
+    #add QCacheHitPct
+    query_cache_hit_pct = Hash.new()
+    qcache_hits = get_value(global_stats,"Qcache_hits")
+    com_select = get_value(global_stats,"Com_select")
+    value = 0
+    if (qcache_hits.to_i + com_select.to_i) != 0
+      value = qcache_hits.to_i/(qcache_hits.to_i + com_select.to_i)*100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: same here to_f

File: source/code/plugins/mysql_workload_lib.rb [hunk 18]
```diff
@@ -0,0 +1,442 @@
...
+    end
+    query_cache_hit_pct.store("QCacheHitPct", value)
+    final_records.push(query_cache_hit_pct)
+
+    #add QCachePrunesPct
+    query_cache_prunes_pct = Hash.new()
+    qcache_prunes = get_value(global_stats,"Qcache_lowmem_prunes")
+    value = 0
+    if queries.to_i != 0
+      value = (qcache_prunes.to_i/queries.to_i) * 100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: same here to_f

File: source/code/plugins/mysql_workload_lib.rb [hunk 19]
```diff
@@ -0,0 +1,442 @@
...
+    query_cache_prunes_pct.store("QCachePrunesPct", value)
+    final_records.push(query_cache_prunes_pct)
+
+    #add TCacheHitPct
+    table_hit_pct = Hash.new()
+    open_tables = get_value(global_stats,"Open_tables")
+    opened_tables = get_value(global_stats,"Opened_tables")
+    value = 0
+    if opened_tables.to_i != 0
+      value = open_tables.to_i/opened_tables.to_i * 100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: same here to_f

File: source/code/plugins/mysql_workload_lib.rb [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+    table_hit_pct.store("TCacheHitPct", value)
+    final_records.push(table_hit_pct)
+
+    #add TableLockContentionPct
+    table_lock_pct = Hash.new()
+    table_lock_waited = get_value(global_stats,"Table_locks_waited")
+    table_lock_immediate = get_value(global_stats,"Table_locks_immediate")
+    value = 0
+    if (table_lock_waited.to_i + table_lock_immediate.to_i) != 0
+      value = table_lock_waited.to_i/(table_lock_waited.to_i + table_lock_immediate.to_i) * 100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: same here to_f

File: source/code/plugins/mysql_workload_lib.rb [hunk 21]
```diff
@@ -0,0 +1,442 @@
...
+    table_lock_pct.store("TableLockContentionPct", value)
+    final_records.push(table_lock_pct)
+
+    #add IDB_BP_HitPct
+    idb_hit_pct = Hash.new()
+    innodb_buffer_pool_reads = get_value(global_stats,"Innodb_buffer_pool_reads")
+    innodb_buffer_pool_read_requests = get_value(global_stats,"Innodb_buffer_pool_read_requests")
+    value = 0
+    if (innodb_buffer_pool_reads.to_i + innodb_buffer_pool_read_requests.to_i) != 0
+      value = innodb_buffer_pool_reads.to_i/(innodb_buffer_pool_reads.to_i + innodb_buffer_pool_read_requests.to_i) * 100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: same here

File: source/code/plugins/mysql_workload_lib.rb [hunk 22]
```diff
@@ -0,0 +1,442 @@
...
+    idb_hit_pct.store("IDB_BP_HitPct", value)
+    final_records.push(idb_hit_pct)
+
+    #add IDB_BP_UsePct
+    idb_use_pct = Hash.new()
+    innodb_buffer_pool_pages_data = get_value(global_stats,"Innodb_buffer_pool_pages_data")
+    innodb_buffer_pool_pages_total = get_value(global_stats,"Innodb_buffer_pool_pages_total")
+    value = 0
+    if( innodb_buffer_pool_pages_total.to_i != 0)
+      value = innodb_buffer_pool_pages_data.to_i/innodb_buffer_pool_pages_total.to_i * 100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: same here to_f

File: source/code/plugins/mysql_workload_lib.rb [hunk 23]
```diff
@@ -0,0 +1,442 @@
...
+    handler_read_first = get_value(global_stats,"Handler_read_first")
+    handler_read_key = get_value(global_stats,"Handler_read_key")
+    handler_read_next = get_value(global_stats,"Handler_read_next")
+    handler_read_prev = get_value(global_stats,"Handler_read_prev")
+    handler_read_rnd_next = get_value(global_stats,"Handler_read_rnd_next")
+    full_scan_reads = handler_read_rnd.to_i + handler_read_rnd_next.to_i
+    all_row_access  = handler_read_rnd.to_i + handler_read_first.to_i + handler_read_key.to_i + handler_read_next.to_i + handler_read_prev.to_i + handler_read_rnd_next.to_i
+    value = 0 #reset
+    if all_row_access != 0
+      value = full_scan_reads/all_row_access * 100
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: same here to_f

File: source/code/plugins/mysql_workload_lib.rb [hunk 24]
```diff
@@ -0,0 +1,442 @@
...
+    final_records = Array.new
+    timestamp = OMS::Common.format_time(time)
+
+    # get mysql global stats and variables
+    mysqlserver_instances = Array.new
+    server_stats = get_server_stats
+    server_stats.each do |row|
+      row.each {
+        |key, value|
+        mysql_server_counters = Hash.new
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: use singular "counter" will be better here

File: source/code/plugins/mysql_workload_lib.rb [hunk 25]
```diff
@@ -0,0 +1,442 @@
...
+  end
+
+  # Convert data to oms format
+  def enumerate (time)
+    #final list of instances to send to oms filtered based on the specific counters supporeted by mysql plugin (per mysql spec)
+    final_records = Array.new
+    timestamp = OMS::Common.format_time(time)
+
+    # get mysql global stats and variables
+    mysqlserver_instances = Array.new
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: can you name it "counters"

File: source/code/plugins/mysql_workload_lib.rb [hunk 26]
```diff
@@ -0,0 +1,442 @@
...
+    mysql_instance["InstanceName"] = mysql_instance_name
+    mysql_instance["Collections"] = mysqlserver_instances
+    #add the instance to the list
+    final_records.push(mysql_instance)
+
+    #get mysql database tables info
+    result_dbs = get_server_databases(@query_get_database)
+
+    result_dbs.each do |row|
+      mysql_db_instances = Array.new
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: counters

File: source/code/plugins/mysql_workload_lib.rb [hunk 27]
```diff
@@ -0,0 +1,442 @@
...
+      db_instance = Hash.new
+      db_instance["Timestamp"] = timestamp
+      db_instance["Host"] = host_name
+      db_instance["ObjectName"]="MySQL Server Database"
+      db_instance["InstanceName"] = row["DatabaseName"]
+      row.delete("DatabaseName") # leave only table name and size columns and remove database name
+      row.each {
+        |key, value|
+        db_counters = Hash.new
```
Comment (robbiezhang) [inline source/code/plugins/mysql_workload_lib.rb]: counter

File: test/code/plugins/mysql_workload_lib_test.rb [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+      @query_global_status_mock = "SHOW GLOBAL STATUS"
+      @query_variables_mock = "SHOW VARIABLES"
+      @query_hostname_mock = "SHOW VARIABLES LIKE 'hostname'"
+      @query_sizeof_all_dbs_mock = 'select "Total" as "Database", SUM(a.data_length) + SUM(a.index_length) as "Size (Bytes)" from information_schema.schemata b left join information_schema.tables a on b.schema_name = a.table_schema'
+      @query_get_databases_mock = 'select b.schema_name as "DatabaseName",COUNT(a.table_name) as "NumberOfTables", SUM(a.data_length) + SUM(a.index_length) as "SizeInBytes" from information_schema.schemata b left join information_schema.tables a on b.schema_name = a.table_schema group by b.schema_name'
+      #define different mock objects used with query function
+      @global_status_mock_records = [{"Variable_name" => "Threads_connected", "Value" => "1"}, {"Variable_name" => "Aborted_connects", "Value" => "0"}, {"Variable_name" => "Uptime", "Value" => "54"},
+        {"Variable_name" => "Key_reads", "Value" => "0"}, {"Variable_name" => "Key_read_requests", "Value" => "0"}, {"Variable_name" => "Queries", "Value" => "1"},
```
Comment (robbiezhang) [inline test/code/plugins/mysql_workload_lib_test.rb]: please give some valid number to verify the to_f/to_i scenario

**Comment by jeffaco** [comment]
As Robbie says, be sure to rebase to master. You're picking up changes that you shouldn't here.

Also, please squash your commits into a single commit message. You don't need three separate commits for this capability.

Robbie is doing a great job reviewing the .rb files. I'll now go look at other files for correctness.

**Comment by robbiezhang** [comment]
Please merge your commits into one. You can use git rebase -i to do so.
I'm looking for a next iteration.

File: build/buildRuby.sh [hunk 29]
```diff
@@ -202,6 +202,9 @@ export PATH=${RUBY_DESTDIR}/bin:$PATH
 echo "Installing Bundler into Ruby ..."
 elevate ${RUBY_DESTDIR}/bin/gem install ${BASE_DIR}/source/ext/gems/bundler-1.10.6.gem
 
+echo "Installig Mysql2 into Ruby ..."
```
Comment (jeffaco) [inline build/buildRuby.sh]: Slightly more descriptive, like:

echo "Installing MySQL2 gem into Ruby ..."

please.

File: docs/How to setup mysql plugin.md [hunk 30]
```diff
@@ -0,0 +1,9 @@
...
+
+How to enable mySQL plugin to collect mySQL server stats and perf data
+* Install mySQL server on the machine
+* Machine has to have libmysqlclient.so installed(can be downloaded and installed through mysql-devel)
```
Comment (jeffaco) [inline docs/How to setup mysql plugin.md]: Is this for customers, or for us? This is VERY brief, and I'd like to see more info here. HOW do you install it on RPM based systems? (yum) How do you install it on DPKG systems (apt-get).

Also, getting libmysqlclient.so is a bit harder on Ubuntu, so you might want to mention the kit to include from apt-get to grab this.

File: installer/datafiles/base_omsagent.data (hunk 3 above)
Comment (jeffaco) [inline installer/datafiles/base_omsagent.data]: Yeah, this is just plain wrong, and I'm surprised you didn't catch this. I'll need another iteration please before signoff on this alone.

File: installer/datafiles/base_omsagent.data [hunk 31]
```diff
@@ -81,8 +81,8 @@ MAINTAINER:              'Microsoft Corporation'
 /opt/microsoft/omsagent/plugin/out_oms_statsd_aggregator.rb;            source/code/plugins/out_oms_statsd_aggregator.rb;      744; root; root
 /opt/microsoft/omsagent/plugin/auditlog_lib.rb;                         source/code/plugins/auditlog_lib.rb;                   744; root; root
 /opt/microsoft/omsagent/plugin/parser_auditlog.rb;                      source/code/plugins/parser_auditlog.rb;                744; root; root
-/opt/microsoft/omsagent/plugin/mongostat_lib.rb;                        source/code/plugins/mongostat_lib.rb;                  744; root; root
-/opt/microsoft/omsagent/plugin/in_mongostat.rb;                         source/code/plugins/in_mongostat.rb;                   744; root; root
+/opt/microsoft/omsagent/plugin/mysql_workload_lib.rb;                   source/code/plugins/mysql_workload_lib.rb;             744; root; root
+/opt/microsoft/omsagent/plugin/in_mysql_workload.rb;                    source/code/plugins/in_mysql_workload.rb;              744; root; root
```
Comment (jeffaco) [inline installer/datafiles/base_omsagent.data]: These are wrong too (because you are not up to date with the master branch). Rebase to master, then send a new iteration.

File: installer/datafiles/ruby.data [hunk 32]
```diff
@@ -3,38 +3,14 @@ RUBY_DEST:                '/opt/microsoft/omsagent/ruby'
 
 %Files
```
Comment (jeffaco) [inline installer/datafiles/ruby.data]: There are a lot of harmless changes here. Some of these might be caused by sorting files in genRubyInstaller.sh. I won't review this, but will review genRubyInstaller.sh.

Did you diff this resulting file with the file I got by running genRubyInstaller on a non-universal system? You should do that, and I saved the file in case you want it (rather than regenerating yourself).

Given that I can't really review ruby.data, we need to be sure that there are no unintended changes here.

**Comment by jeffaco** [comment]
You didn't do my suggested changes to genRubyInstaller.sh, and even worse, you didn't include documentation (in .md form) of how others can add components to Ruby. I won't sign off on this until I see both of those please, thanks. The documentation is very important for your group to be able to support this down the road, and not to be so reliant on me.

Thanks for your understanding.

File: build/buildRuby.sh (hunk 29 above)
Comment (jeffaco) [inline build/buildRuby.sh]: Also, you changed the build process for Ruby, but you did not bump the directory in /usr/local. This will break unit tests. If you change this file, it's highly likely that you need to bump the directory. In your case, it's mandatory.

File: installer/datafiles/ruby.data (hunk 32 above)
Comment (NarineM) [inline installer/datafiles/ruby.data]: The file you got by running your private version of genRubyInstaller has no changes whatsoever compared to the base line. Those changes are purely content re-formatting related. In fact running the genRubyInstaller does not give anything as compared to the new changes for ruby. Because even if the new gem is installed from the intermediate or build location , the gem gets binplaced under /opt/Microsoft path by default. I am going to document what are my findings and how I got the new changes of ruby.data. Please bare with me for the new code review  specifically for ruby installer related fixes.

Comment (NarineM) [inline installer/datafiles/ruby.data]: Yes, Jeff, I did not include ruby installer changes here on purpose. These are not related to MySQL plugin feature work. As noted in the comments section, I  will be sending a separate code review for the ruby installer script and its documentation. Thank you for your review and feedback . Much appreciated.

File: docs/How to setup mysql plugin.md (hunk 1 above)
Comment (NarineM) [inline docs/How to setup mysql plugin.md]: Ok. Fixed. Thanks.

File: installer/datafiles/base_omsagent.data (hunk 3 above)
Comment (NarineM) [inline installer/datafiles/base_omsagent.data]: Yes, I saw it too . Sorry hit the request button too quick . Will have it fixed in the next iteration. Thanks.

File: source/code/plugins/in_mysql_workload.rb (hunk 5 above)
Comment (NarineM) [inline source/code/plugins/in_mysql_workload.rb]: ok.

File: source/code/plugins/in_mysql_workload.rb (hunk 6 above)
Comment (NarineM) [inline source/code/plugins/in_mysql_workload.rb]: Will fix it.

File: source/code/plugins/in_mysql_workload.rb (hunk 7 above)
Comment (NarineM) [inline source/code/plugins/in_mysql_workload.rb]: Fixed.

File: source/code/plugins/mysql_workload_lib.rb (hunk 10 above)
Comment (NarineM) [inline source/code/plugins/mysql_workload_lib.rb]: Fixed the comments formatting.

File: source/code/plugins/mysql_workload_lib.rb (hunk 12 above)
Comment (NarineM) [inline source/code/plugins/mysql_workload_lib.rb]: It is actually always only one value. Each loop here is not needed at all. Will remove the each.

File: source/code/plugins/mysql_workload_lib.rb (hunk 13 above)
Comment (NarineM) [inline source/code/plugins/mysql_workload_lib.rb]: There is always one property with the name hostname.

File: source/code/plugins/mysql_workload_lib.rb (hunk 14 above)
Comment (NarineM) [inline source/code/plugins/mysql_workload_lib.rb]: Yes, there are case sensitive. the names from environment queries are lower case when global status returns upper case.

File: source/code/plugins/mysql_workload_lib.rb (hunk 15 above)
Comment (NarineM) [inline source/code/plugins/mysql_workload_lib.rb]: Ah, good catch!

File: source/code/plugins/mysql_workload_lib.rb (hunk 24 above)
Comment (NarineM) [inline source/code/plugins/mysql_workload_lib.rb]: Fixed. Thanks.

File: test/code/plugins/mysql_workload_lib_test.rb (hunk 28 above)
Comment (NarineM) [inline test/code/plugins/mysql_workload_lib_test.rb]: Adding a use case for to_f.

File: docs/How to setup mysql plugin.md (hunk 30 above)
Comment (NarineM) [inline docs/How to setup mysql plugin.md]: This is a temp content. Please see the comments/description section of this pull request. We already have MySQL documented for end user that will need to be changed once the omi provider is removed. For now, I am just adding a few lines as readme. Will add more details if you find it useful.

File: installer/datafiles/base_omsagent.data (hunk 31 above)
Comment (NarineM) [inline installer/datafiles/base_omsagent.data]: Ah, that is my bad. Hit the send button too quick :)

File: build/buildRuby.sh (hunk 29 above)
Comment (NarineM) [inline build/buildRuby.sh]: The change for the test version of ruby is in build/configure. Somehow missed that file. Will add in the next iteration. Good catch!

File: installer/conf/omsagent.d/mysqlworkload.conf (hunk 2 above)
Comment (NarineM) [inline installer/conf/omsagent.d/mysqlworkload.conf]: separate the words by underscore or dashes? Any recommendation regarding the conf file name?

File: installer/datafiles/ruby.data (hunk 32 above)
Comment (jeffaco) [inline installer/datafiles/ruby.data]: Hi Narine,

What you said is not correct. The new ruby.data file that you generated absolutely has changes, and has changes that you need. For example, one of the changes is:

${{RUBY_DEST}}/lib/ruby/gems/2.2.0/doc/mysql2-0.4.4/ri/ext/mysql2/page-Makefile.ri; ${{RUBY_INT}}/lib/ruby/gems/2.2.0/doc/mysql2-0.4.4/ri/ext/mysql2/page-Makefile.ri; 644; root; root

Either check in the changes to genRubyInstaller.sh first (along with documentation on how to update Ruby), or include those files in this pull request. I don't care which. But if you needed to update genRubyInstaller.sh as part of this work, then that should either be checked in as part of this work, or checked in first so it's clearer how this work was done.

I'm perfectly fine helping your team, but it's super important to me that we improve documentation to ease my workload going forward. Thus, as far as I'm concerned, my help to you to complete this work is actually just as important as improving documentation so, next time this comes up, somebody can just look at the docs.

Also, since we need to update genRubyInstaller.sh anyway, it's important to me that that is done properly so that, after this commit, repositioning changes in ruby.data go away. We need to eat that once (as part of this checkin), but we shouldn't need to eat that twice due to incorrect changes to genRubyInstaller.sh.

With the suggested changes to genRubyInstaller, future changes to Ruby will be much easier to understand from the file diff.

Thanks for your understanding. I consider this work mandatory for my signoff to this PR. You can get genRubyInstaller.sh (and documentation additions) checked in first, if you want, or you can include it in this PR. But I won't sign off on this one without seeing that first, and knowing that the messy ruby.data diffs are just a one-time thing.

Thanks for your understanding.

File: build/buildRuby.sh (hunk 29 above)
Comment (NarineM) [inline build/buildRuby.sh]: Added the word "gem" to the echo message.

File: installer/datafiles/ruby.data (hunk 32 above)
Comment (NarineM) [inline installer/datafiles/ruby.data]: @jeffaco 
The data file that I generated has the changes, that is correct, but I have not run the script with the SOURCE dir pointing to intermediate ruby paths at all. That is what I think is broken in the version of genRubyInstaller .

File: source/code/plugins/in_mysql_workload.rb (hunk 9 above)
Comment (NarineM) [inline source/code/plugins/in_mysql_workload.rb]: connection will eventually time out and be destroyed but I agree that it would be better to add explicit release of the resource. Fixing now.

File: docs/How-to-setup-mysql-plugin.md [hunk 33]
```diff
@@ -0,0 +1,14 @@
+# How to setup Mysql plugin for OMS agent
+
+How to enable mySQL plugin to collect mySQL server stats and perf data
+* Machine has to have libmysqlclient.so installed (which can be downloaded and installed through mysql-devel.
+					yum install mysql-devel or apt-get mysql-devel for Debian systems)
```
Comment (jeffaco) [inline docs/How-to-setup-mysql-plugin.md]: Your instructions are not correct for Debian/Ubuntu, please verify proper command and update.

**Comment by jeffaco** [comment]
I'm still looking for another review. I need to see (either in a prior checkin to this or combined with this):

- Modified genRubyInstaller.sh that works in the current environment without changes,
- Instructions on how to update Ruby in the future, when future changes are necessary.
- Modified Ruby.data file from new genRubyInstaller.sh (which will minimally sort the entries so, in the future, "false changes" are dramatically reduced).

Also, as mentioned in prior review, please squash your commits. You don't need 12 commit entries for what should be 1 (perhaps 2 if you change genRubyInstaller.sh and commit genRubyInstaller.md separately).

As stated earlier, I won't sign off without these changes as well. Thanks.

File: docs/How-to-setup-mysql-plugin.md (hunk 33 above)
Comment (NarineM) [inline docs/How-to-setup-mysql-plugin.md]: Fixed the readme. Most debian systems come with this library. Also installing through MySQL-devel would not be always used as an option in production environment. So I have updated the readme to have just one example for installing it through devel package.

**Comment by NarineM** [comment]
@jeffaco 
Thanks Jeff for your review. I am going to send a seperate code review for genRubyInstaller and it is documentation since these are for future use and are not impacting this pull request. Bullet#3(updating ruby.data to reduce diffs for future changes) and squashing commits are already done. See the latest commit. Thanks.

**Comment by robbiezhang** [comment]
Signoff on the business logic. looking for the bundle part

**Comment by jeffaco** [comment]
Hi Narine,

There's a very good chance that there will be further updates to Ruby.data, depending on your changes to genRubyInstaller.sh. That's why I wanted this part done before or as part of this checkin. But you committed anyway without signoff for me, which I think is incorrect process. But I'll leave that up to Robbie to handle (either roll back your changes and update this, or let is go).

In the future, if someone says they do not sign off, you MUST reconcile with them before committing. At least that's the way my group works.

If Robbie decides not to roll back your commit, when can I expect a PR to address the rest of this? By EOD on Tuesday?

**Comment by NarineM** [comment]
Hi Jeff,
I will have the changes for genRubyInstaller by EOD Wed. This PR did not intend to include the change for the installer script. I will send a separate PR for the installation piece and its documentation altogether.
